---
title: "scRNAseq Analysis"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: "hide"
    theme: "cosmo"
    highlight: "pygments"
    fig_width: 7
    fig_height: 5
    css: styles.css
  pdf_document:
    toc: true
    toc_depth: 2
    number_sections: true
    highlight: "pygments"
    latex_engine: xelatex
    extra_dependencies: ["caption"]
    fig_width: 7
    fig_height: 5
params:
  file_10X_h5: NULL
fontsize: 11pt
geometry: margin=1in
---



```{css, echo=FALSE}
body {
  font-family: Arial, sans-serif;
  line-height: 1.6;
}
.text-box {
  border: 1px solid #dcdcdc;
  padding: 15px;
  margin: 20px 0;
  background-color: #e6f7ff; /* Light blue background color */
  border-radius: 5px;
}
.title-box {
  background-color: #007acc;
  color: white;
  padding: 10px;
  border-radius: 5px;
  font-size: 1.5em;
  margin-bottom: 20px;
}

# Create the "plots" directory if it doesn't exist

  
```

***

scRNAseq Analysis Pipeline


# Overview

<div class="text-box">
This document outlines an automated pipeline for analyzing single-cell RNA sequencing (scRNA-seq) data using the 10x Genomics Chromium platform with 5' gene expression (GEX) captures. The analysis includes>
1. Loading data from `.h5` files

2. Performing quality control and normalization
    
3. Cell cycle scoring
    
4. Annotating cells using SingleR

5. Integrating multiple samples

6. Visualizing results

7. Identifying marker genes

8. Differential expression and marker selection

9. Cell type annotation using SingleR
</div>



```{r setup, include=FALSE}
# Load required libraries
library(xfun)
library(dplyr)
library(hdf5r)
library(Seurat)
library(ggplot2)
library(ggraph)
library(RColorBrewer)
library(data.table)
library(SingleR)
library(celldex)
library(knitr)
knitr::opts_chunk$set(dev = 'png')
```



# 0. Load the data

```{r}
# Function to load data and create Seurat object
load_and_create_seurat_object <- function(file_path, project_name) {
  data <- Read10X_h5(file_path, use.names = TRUE, unique.features = TRUE)
  
  # Check and extract 'Gene Expression' modality if present
  if (is.list(data) && "Gene Expression" %in% names(data)) {
    rna_data <- data$`Gene Expression`
  } else if (is.list(data)) {
    rna_data <- data[[1]]
  } else {
    rna_data <- data
  }

  # Create Seurat object
  seurat_object <- CreateSeuratObject(counts = rna_data, project = project_name, min.cells = 3, min.features = 100)
  
  return(seurat_object)
}

# Load either PBMC or Tumor data
file_path <- "/home/mmerono/Downloads/sc5p_v2_hs_PBMC_10k_filtered_feature_bc_matrix.h5"
project_name <- "scRNAseq_analysis"  
seurat_obj <- load_and_create_seurat_object(file_path, project_name)

```

# 1. Quality control (QC) and filtering

### Compute QC Metrics

```{r }
# Add number of genes per UMI for each cell to metadata
seurat_obj$log10GenesPerUMI <- log10(seurat_obj$nFeature_RNA) / log10(seurat_obj$nCount_RNA)

# Compute percent mito ratio
seurat_obj$mitoRatio <- PercentageFeatureSet(object = seurat_obj, pattern = "^MT-") / 100

```

### Ratio of mitochondiral

In single-cell RNA sequencing analysis, we aim to filter out specific cells to ensure data quality. First, we remove cells with a low number of genes or total molecules, as these potentially represent low-quality or empty cells, which can introduce noise into the dataset. Next, we exclude cells with an abnormally high number of genes or total molecules because these may be doublets or multiplets, where two or more cells were captured together, introducing bias to the data. Lastly, we filter out cells with a high percentage of mitochondrial genes, as this often indicates dying or low-quality cells, which can compromise the accuracy and reliability of the analysis.

```{r }
# Compute percent mito ratio
seurat_obj$mitoRatio <- PercentageFeatureSet(object = seurat_obj, pattern = "^MT-")
seurat_obj$mitoRatio <- seurat_obj@meta.data$mitoRatio / 100

```

```{r }
# Create metadata dataframe
metadata <- seurat_obj@meta.data

# Add cell IDs to metadata
metadata$cells <- rownames(metadata)

# Rename columns
metadata <- metadata %>%
  dplyr::rename(seq_folder = orig.ident, nUMI = nCount_RNA, nGene = nFeature_RNA)

# Add metadata back to Seurat object
seurat_obj@meta.data <- metadata

                           
```

### Cell counts visualization

```{r, fig.align="center", fig.width=6, fig.height=6, fig.cap="Figure 1: ." }

# Visualize the correlation between genes detected and number of UMIs
p1 <- metadata %>% 
  ggplot(aes(x = nUMI, y = nGene, color = mitoRatio)) + 
  geom_point() + 
  scale_colour_gradient(low = "gray90", high = "black") +
  stat_smooth(method = "lm", color = "blue") +
  scale_x_log10() + 
  scale_y_log10() + 
  theme_classic() +
  geom_vline(xintercept = 500, linetype = "dashed") +
  geom_hline(yintercept = 250, linetype = "dashed") +
  labs(
    title = "Correlation between Genes Detected and Number of UMIs",
    x = "Number of UMIs (log scale)",
    y = "Number of Genes (log scale)",
    color = "Mitochondrial Ratio"
  ) +
  theme(plot.title = element_text(hjust = 0.5))

# Save the plot
ggsave("plots/cell_counts_visualization.png", plot = p1)
p1


```



In the context of quality control (QC) metrics visualization in single-cell RNA sequencing analysis, we employ filtering criteria to refine the dataset. Firstly, we target cells with unique feature counts exceeding 2,500 or falling below 200, as these extremes may indicate technical artefacts or low-quality cells. Additionally, we identify cells with mitochondrial counts surpassing 5%, indicative of potential cell stress or poor quality, and subsequently remove them from the dataset to mitigate their adverse effects on data interpretation. 

###Violin plot of QC metrics

```{r, fig.align="center", fig.width=6, fig.height=6, fig.cap="Figure 2: ."}
# Visualize QC metrics as a violin plot
p2 <- VlnPlot(
  object = seurat_obj,
  features = c("nUMI", "nGene", "mitoRatio"),
  pt.size = 0.01,
  ncol = 3
) & 
  theme(
    axis.text.x = element_text(angle = 0, hjust = 0.5),
    axis.title.x = element_blank()
  )

# Save the plot
 ggsave("plots/qc_violin_plot.png", plot = p2)

p2


```

Cell-level filtering

-nUMI > 1000

-nGene > 500

-log10GenesPerUMI > 0.8

-mitoRatio < 0.2


```{r }
# Filter out low-quality cells using selected thresholds
filtered_pbmc <- subset(
  x = seurat_obj,
  subset = (nUMI >= 500) & 
    (nGene >= 250) & 
    (log10GenesPerUMI > 0.80) & 
    (mitoRatio < 0.20)
)
			                          
```

# 2. Normalizing the data

After removing unwanted cells from the dataset, the next step is to normalize the data.


```{r }
# Standard log normalization for RNA
filtered_pbmc <- NormalizeData(filtered_pbmc, normalization.method = "LogNormalize", scale.factor = 10000, assay = "RNA")

```

## Cell cycle scoring

Calculate cell cycle scores. Seurat has build-in list, cc.genes.updated.2019, that defines genes involved in cell cycle. 

```{r }
s.genes <- cc.genes.updated.2019$s.genes
g2m.genes <- cc.genes.updated.2019$g2m.genes

filtered_pbmc <- CellCycleScoring(filtered_pbmc, s.features = s.genes, g2m.features = g2m.genes)
table(filtered_pbmc[[]]$Phase)

```



# 3. Identification of highly variable features (feature selection)

```{r }
filtered_pbmc <- FindVariableFeatures(filtered_pbmc, selection.method = "vst", nfeatures = 2000)
```


# 4. Scaling the data

Remove unwanted sources of variation.

```{r }
# Scale the counts
filtered_pbmc <- ScaleData(filtered_pbmc)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(filtered_pbmc), 10)

# Plot variable features with and without labels
var_features <- VariableFeaturePlot(filtered_pbmc)
lab_points <- LabelPoints(plot = var_features, points = top10, repel = TRUE)

# Save the plots
ggsave("plots/variable_features.png", plot = var_features)
ggsave("plots/variable_features_labeled.png", plot = lab_points)

lab_points

```


# 5. Performing linear dimensional reduction 

Next, we perform PCA on the scaled data.

By default we see 2000 most variable genes.

We also split the figure by cell cycle phase, to evaluate similarities and/or differences.

```{r }
filtered_pbmc <- RunPCA(filtered_pbmc, features = VariableFeatures(object = filtered_pbmc))

# Plot the PCA colored by cell cycle phase
p3 <- DimPlot(filtered_pbmc, reduction = "pca", group.by = "Phase", split.by = "Phase")

# Save the plot
ggsave("plots/pca_phase.png", plot = p3)

p3

```

```{r}
print(filtered_pbmc[["pca"]], dims = 1:5, nfeatures = 5)

p4 <- VizDimLoadings(filtered_pbmc, dims = 1:2, reduction = "pca")

# Save the plot
ggsave("plots/pca_loadings.png", plot = p4)

p4

p5 <- DimPlot(filtered_pbmc, reduction = "pca")

# Save the plot
ggsave("plots/pca_plot.png", plot = p5)

p5

p6 <- DimHeatmap(filtered_pbmc, dims = 1, cells = 500, balanced = TRUE)

# Save the plot
ggsave("plots/pca_heatmap1.png", plot = p6)

p6

p7 <- DimHeatmap(filtered_pbmc, dims = 1:15, cells = 500, balanced = TRUE)

# Save the plot
ggsave("plots/pca_heatmap2.png", plot = p7)

p7

```


### Determine the dimensionality of the dataset


To determine the number of significant principal components (PCs):

```{r }
p8 <- ElbowPlot(filtered_pbmc)

# Save the plot
ggsave("plots/elbow_plot.png", plot = p8)

p8

```


Where the "elbow" appears is usually the threshold for identifying the majority of the variation. This method can be a bit subjective about where to locate the "elbow". To help identifying we can use this plot:


```{r }
# Determine percent of variation associated with each PC
pct <- filtered_pbmc[["pca"]]@stdev / sum(filtered_pbmc[["pca"]]@stdev) * 100

# Calculate cumulative percents for each PC
cumu <- cumsum(pct)

# Determine which PC exhibits cumulative percent greater than 90% and % variation associated with the PC as less than 5
co1 <- which(cumu > 90 & pct < 5)[1]

# Determine the difference between variation of PC and subsequent PC
co2 <- sort(which((pct[1:length(pct) - 1] - pct[2:length(pct)]) > 0.1), decreasing = TRUE)[1] + 1

# Minimum of the two calculations
pcs <- min(co1, co2)

# Create a dataframe with values
plot_df <- data.frame(pct = pct, cumu = cumu, rank = 1:length(pct))
opt_dims <- plot_df$rank[plot_df$rank > pcs][1]

# Elbow plot to visualize
p9 <- ggplot(plot_df, aes(cumu, pct, label = rank, color = rank > pcs)) + 
  geom_text() + 
  geom_vline(xintercept = 90, color = "grey") + 
  geom_hline(yintercept = min(pct[pct > 5]), color = "grey") +
  theme_bw()

# Save the plot
ggsave("plots/elbow_plot_annotated.png", plot = p9)

p9



```

*Developed by members of the teaching team at the Harvard Chan Bioinformatics Core (HBC).*

*Seurat recommends a default resolution of 0.8 for typical single-cell datasets. A higher resolution may be more suitable for larger datasets and vice versa.*

# 6. Cluster the cells

```{r }
filtered_pbmc <- FindNeighbors(filtered_pbmc, dims = 1:opt_dims)
filtered_pbmc <- FindClusters(filtered_pbmc, resolution = 0.8)

# View cluster IDs of the first 5 cells
head(Idents(filtered_pbmc), 5)

p10 <- DimPlot(filtered_pbmc, group.by = "RNA_snn_res.0.8", label = TRUE)

# Save the plot
ggsave("plots/cluster_plot.png", plot = p10)

p10


```


The lower the resolution the less clusters it will group the data in. High resolution will try to find more clusters. 
Cells are clustered based on their gene expression profile. 

Since gene expression of the basis of differentiation, different clusters usually represents different cell types. 

```{r }
# Look at cluster IDs of the first 5 cells
head(Idents(filtered_pbmc), 5)

```

# 7. Non-linear dimantisonal reduction (tSNE/UMAP)

By default, the clusters are numbered by the number of cells with cluster 0 having the largest number of cells. 

## umap 

```{r }
filtered_pbmc <- RunUMAP(filtered_pbmc, dims = 1:opt_dims)
p11 <- DimPlot(filtered_pbmc, reduction = "umap", label = TRUE)

# Save the plot
ggsave("plots/umap_plot.png", plot = p11)

p11

```

## tsne 

```{r }
filtered_pbmc <- RunTSNE(filtered_pbmc, dims = 1:20) 
p12 <- TSNEPlot(object = filtered_pbmc, label = TRUE)

# Save the plot
ggsave("plots/tsne_plot.png", plot = p12)

p12

```
*Done following SANGER singel cell workflow.*

```{r }
# Explore whether clusters segregate by cell cycle phase
p13 <- DimPlot(filtered_pbmc, label = TRUE, split.by = "Phase") + NoLegend()

# Save the plot
ggsave("plots/phase_seg_plot.png", plot = p13)

p13


```

## Cluster segregation by QC metrics

The parameter min.cutoff of q10 is translated as 10% of the cells with the lowest expression of the gene will not exhibit any purple shading. This parameter is applied because when plotting, the order of the arguments will plot the positive cells above the negative cells and with min.cutoff argument we can determine the threshold for shading. 

```{r }

metrics <-  c("nUMI", "nGene", "S.Score", "G2M.Score", "mitoRatio")

p14 <- FeaturePlot(filtered_pbmc, 
            reduction = "umap", 
            features = metrics,
            pt.size = 0.4, 
            order = TRUE,
            min.cutoff = 'q10',
            label = TRUE)

# Save the plot
ggsave("plots/feature_plot.png", plot = p14)

p14

```


*Done following hbctraining workflow.*

##Boxplot of nGene per cluster

```{r }
p15 <- ggplot(filtered_pbmc@meta.data) +
  geom_boxplot(aes(x = RNA_snn_res.0.8, y = nGene, fill = RNA_snn_res.0.8)) +
  NoLegend()

# Save the plot
ggsave("plots/nGene_boxplot.png", plot = p15)

p15

```



# 8. Differential expression and marker selection

Differential gene expression allows us to define gene markers specific to each cluster.


```{r }
DefaultAssay(filtered_pbmc) <- "RNA"
filtered_pbmc <- NormalizeData(filtered_pbmc)
filtered_pbmc <- FindVariableFeatures(filtered_pbmc, selection.method = "vst", nfeatures = 2000)
all.genes <- rownames(filtered_pbmc)
filtered_pbmc <- ScaleData(filtered_pbmc, features = all.genes)

```


Find markers for every cluster by comparing it to all remaining cells, while reporting only the positive ones. The test used: Wilcoxon Rank Sum. 


```{r }
all.markers <- FindAllMarkers(filtered_pbmc, only.pos = TRUE, min.pct = 0.5, logfc.threshold = 0.5)
```


```{r }
table(all.markers$cluster)

top3_markers <- as.data.frame(all.markers %>% group_by(cluster) %>% top_n(n = 3, wt = avg_log2FC))
top3_markers

```


```{r }
top3_markers <- as.data.frame(all.markers %>% group_by(cluster) %>% top_n(n = 3, wt = avg_log2FC))
top3_markers
```

# 9. Cell type annotation using SingleR

We can try automatic annotation with SingleR, using reference dataset from celldex package

```{r}
monaco.ref <- celldex::MonacoImmuneData()
```

convert our Seurat object to single cell experiment for convinience:

```{r }
sce <- as.SingleCellExperiment(DietSeurat(filtered_pbmc))
sce
```


```{r }
monaco.main <- SingleR(test = sce,assay.type.test = 1,ref = monaco.ref,labels = monaco.ref$label.main)
monaco.fine <- SingleR(test = sce,assay.type.test = 1,ref = monaco.ref,labels = monaco.ref$label.fine)

```


```{r }
table(monaco.main$pruned.labels)
table(monaco.fine$pruned.labels)

```


Add the annotations to the Seurat Object metadata.

```{r }
filtered_pbmc@meta.data$monaco.main <- monaco.main$pruned.labels
filtered_pbmc@meta.data$monaco.fine <- monaco.fine$pruned.labels
```

```{r }
srat <- SetIdent(filtered_pbmc, value = "monaco.fine")
p16 <- DimPlot(srat, label = TRUE, repel = TRUE, label.size = 3) + NoLegend()

# Save the plot
ggsave("plots/annotation_plot.png", plot = p16)

p16

```



Extract top 10 cell types and their top 10 expressed genes

```{r }

# Get the top 10 cell types
top_cell_types <- names(sort(table(filtered_pbmc@meta.data$monaco.main), decreasing = TRUE))[1:10]

# Find top 10 expressed genes for each cell type
top_genes <- lapply(top_cell_types, function(cell_type) {
  cells <- which(filtered_pbmc@meta.data$monaco.main == cell_type)
  avg_exp <- rowMeans(GetAssayData(filtered_pbmc, slot = "data")[, cells])
  names(sort(avg_exp, decreasing = TRUE))[1:10]
})

# Combine the list of top genes
top_genes <- unique(unlist(top_genes))

# Create dot plot based on RNA expression
p17 <- DotPlot(filtered_pbmc, features = top_genes, assay = "RNA") +
  ggtitle("Top 10 Cell Types and Their Top 10 Expressed Genes") +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

# Save the plot
ggsave("plots/dot_plot_top_genes.png", plot = p17)

# Display the plot
print(p17)


```





## Differential expression analysis

```{r }

# Find markers
oupMarker <- FindAllMarkers(filtered_pbmc, only.pos = TRUE, logfc.threshold = 1.0, min.pct = 0.2)
oupMarker <- data.table(oupMarker)
oupMarker$pct.diff = oupMarker$pct.1 - oupMarker$pct.2
oupMarker <- oupMarker[, c("cluster","gene","avg_log2FC","pct.1","pct.2","pct.diff","p_val","p_val_adj")]

# Check if known genes are in the marker gene list
knownGenes <- c("CD34","CRHBP","GATA1", "CD14","IRF8","CD19","CD4","CD8B","GNLY")
oupMarker[gene %in% knownGenes]

# Get top genes for each cluster and do dot plot / violin plot
oupMarker$cluster = factor(oupMarker$cluster, levels = unique(filtered_pbmc$seurat_clusters))
oupMarker = oupMarker[order(cluster, -avg_log2FC)]
genes.to.plot <- unique(oupMarker[cluster %in% unique(filtered_pbmc$seurat_clusters), head(.SD, 2), by="cluster"]$gene)

# Set color for gene expression
colGEX = c("grey85", brewer.pal(7, "Reds"))

p1 <- DotPlot(filtered_pbmc, group.by = "seurat_clusters", features = genes.to.plot) + 
  coord_flip() + scale_color_gradientn(colors = colGEX) +
  theme(axis.text.x = element_text(angle = -45, hjust = 0))

# Setup color palettes
nClust <- uniqueN(Idents(filtered_pbmc))
colCls <- colorRampPalette(brewer.pal(n = 10, name = "Paired"))(nClust)

p2 <- VlnPlot(filtered_pbmc, group.by = "seurat_clusters", fill.by = "ident", cols = colCls, features = genes.to.plot, stack = TRUE, flip = TRUE)

```

```{r }
p1
p2
```



## Finding differentially expressed features (look for genetic markers)

Seurat function can find markers that define clusters via differential expression. It identifies positive and negative markers of a single cluster, compared to all other cells (using FindMarkers( ident.1 = )). The function "FindAllMarkers" automates this process ofr all clusters, but it can only be tested groups of clusters vs. each other, or against all cells.  
"min.pct" argument requires a feature to be detected at a minimum percentage in either of the two groups pf cells. 
"thresh.test" argument requires a feature to be differentially expressed on average by some amount between two groups. 
"max.cells.per.ident" is an argument used to speed up computations by downsample each identity class to have no more cells than what is set to. 


<!-- Footer -->
<hr />
<p style="text-align: center;">A work by <a href="https://github.com/mmerono/scRNA-seq-pipeline.git">Marta Meroño</a></p>
<p style="text-align: center;"><span style="color: #808080;"><em>mmeronorafel@gmail.com</em></span></p>

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">

<p style="text-align: center;">
    <a href="https://www.linkedin.com/in/yan-holtz-2477534a/" class="fa fa-linkedin"></a>
    <a href="https://github.com/mmerono/scRNA-seq-pipeline.git" class="fa fa-github"></a>
</p>

